### DISCUSSION TITLE
Unmatched UserAgent Version

### DISCUSSION THREAD

**Discussion by KrAsH-CoD3** [kaliiiiiiiiii/Selenium-Profiles#25] [General]
Hi, this project is awesome project. Putting together uc, auth proxy-enabled driver, and other awesome features. **Thank you** for this project.
I am however trying to use another UserAgent in the `useragent["userAgent"]` value I also try to change its version where necessary but [Iphey](https://iphey.com/) seems to see my user agent version that it doesn't match. 

I tried several mobile user agents changing its version also where necessary in the profile.json but the same.

Do you by chance know the factors which account for this mismatch of UA in general?
Here's how my profile looks using an old android device UA:
```json
"useragent": {
                        "platform": "Linux aarch64",
                        "acceptLanguage":"en-US",
                        "userAgent": "Mozilla/5.0 (Linux; Android 7.1.1; GIONEE S10 lite) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/109.0.0.0 Mobile Safari/537.36",
                        "userAgentMetadata": {
                            "brands": [{"brand": "Google Chrome", "version": "109"}, {"brand": "Not)A;Brand", "version": "8"},
                                      {"brand": "Chromium", "version": "109"}],
                            "fullVersionList": [{"brand": "Google Chrome", "version": "109.0.0.0"},
                                                {"brand": "Not)A;Brand", "version": "8.0.0.0"},
                                                {"brand": "Chromium", "version": "109.0.0.0"}],
                            "fullVersion": "109.0.0.0",
                            "platform": "Android",
                            "platformVersion": "7.1.1",
                            "architecture": "",
                            "model": "GIONEE S10 lite",
                            "mobile": true,
                            "bitness": "",
                            "wow64": false}
            }
```

**Comment by ttlns** [comment]
Thanks for your feedback!
1. Does [Iphey](https://iphey.com/) also report a inconsistent __UA-Version__  on the default `profiles.Andoid` profile?
2. It looks to me like you only changed `"userAgent"`, `"version"`, `"fullVersion"` and `"platformVersion"`
    2.1 `"platform"` doesn't always need to be `"Linux aarch64"` I think. I assume it's dependent form `"model"`, `"userAgent"` and `"platformVersion"`
    2.2 I don't think the full version of chrome would ever be `"109.0.0.0"` in reality, maybe that's what [Iphey](https://iphey.com/) detects.  see [example-real-device](https://github.com/kaliiiiiiiiii/Selenium-Profiles/blob/master/src/selenium_profiles/profiles/default.json)
3. If you have access to the device you want to "clone", check what profile https://js.do/kaliiiiiiiiiii/get_profile returns you:) If not, maybe try finding similar devices with 109 chrome-version.

Note: I unfortanerly don't have a lot of example profiles to compare with, for example Linux, Ipad, Tablet, different versions, different devices are missing yet.

**Reply by KrAsH-CoD3** [reply]
Thanks for your swift response. 
`profiles.Andoid` do have a consistent UA-Version. 

- I tried 3 and it worked. I did a little play around. I noticed I always get Unmatch UA-Version whenever `{"brand": "Not)A;Brand"}` of `userAgentMetadata["brands"][1]["brand"]`, the value `Not)A;Brand` is not a valid value in this case. 
- The extracted profile value that work is `"Not_A Brand"`. 
- I tried several values `"Not A(Brand"`, `"Not)A;Brand"`, `"Not A;Brand"`, `"Not)A(Brand"` e.t.c of which [Iphey](https://iphey.com/) displayed as a valid browser despite even if other values of (`"model"`, `"platformVersion"` e.t.c ) were not matched correctly. 
- Intentionally also used unmatched values for `"platformVersion"`, `"fullVersion"` e.t.c and it produced a valid browser still despite using `"Not A(Brand"`, `"Not)A;Brand"`, `"Not A;Brand"`, `"Not)A(Brand"`

Does this mean for any UA we want to use, we have to extract its profile from [get profile](https://js.do/kaliiiiiiiiiii/get_profile) so we get the correct value for each key? The reason for this is cus tried emulated Devices in Chrome but it seems to get Windows properties of the value instead of the emulated device value.

**Reply by ttlns** [reply]
Well usually, websites which are browser-fingerprint protected collect a lot of data, analyse patterns for devices. With that, they detect unusual values.

For now, I'd recommend you to clone devices, as I unfortunately don't have access to a fingerprint database which includes all necessary values.  
Or, if the protection isn't that advanced, you might as well find some patters by yourself to generate different profiles.

And yes, 'normal" device emulation in chrome seems to miss some values.

**Reply by KrAsH-CoD3** [reply]
Thank you.

**Comment by ttlns** [comment]
Selenium-profiles now automatically patches the Chrome/Version 
```python
profile["cdp"]["patch_version"] = True | False | "full_version"
```
for example `114.0.5735.90`)
